src/cli/README_COAPS.md

```diff
@@ -0,0 +1,155 @@
+# OpenThread CLI - COAPS Example
+
+The OpenThread CoAP Secure APIs may be invoked via the OpenThread CLI.
+
+CoAP Secure use DTLS (over UDP) to make a end to end encrypted connection.
+
+## Init (for Server and Client)
+
+ 1. start (init) the coaps api and starts listen on coaps port (5684)
+ 
+```bash
+> coaps start
+```
+
+### For use PSK with AES128 CCM8
+
+ 2a. enter your psk and his identifier
+ 
+```bash
+coaps set psk <yourPsk> <PskIdentifier>
+```
+
+### For use ECDHE ECDSA with AES128 CCM8
+
+ 2b. set the private key and .X509 certificate stored in core/cli/x509_cert_key.hpp.
+ 
+ > _optional_: add your [own](#create-ec-private-key) X.509 certificate and private key to 'core/cli/x509_cert_key.hpp'.
+ 
+```bash
+coaps set x509
+```
+
+## CoAP Secure Server
+
+ * add a coap resource to provide to a coap client
+ 
+```bash
+coaps resource <coapUri>
+```
+
+## Connect to a DTLS Server
+
+ * connect to a dtls server
+ 
+```bash
+coaps connect <serversIp>
+```
+
+## CoAP Secure Client
+
+* get a resource form the server
+
+```bash
+coaps get (serversIp) <coapUri> (Con/NotCon) (payload)
+```
+
+> post, put and delete also possible
+   
+   
+<> must
+() opt
+   
+   
+## Complete example for DTLS/CoAP server (Node 1)
+
+In this example the coap server is also the dtls server.
+The dtls server waits for incoming connection on coaps port 5684.
+The [Node 2](#complete-example-dtlscoap-client-node-2) below is able to connect to this coaps server. 
+
+> Note: Node 1 and Node 2 must use the same mode. Either PSK or Certificate based.   
+
+```
+ Node 1
+---------
+|CoAPS  |
+|Server | <--Listen on Port 5684-- (Node 2)
+|       |
+---------
+```
+
+### with PSK
+
+```bash
+coaps start
+coaps set psk secretPSK Client_identity
+coaps resource time
+```
+
+### with Certificate
+
+```bash
+coaps start (false)
+coaps set x509
+coaps resource time
+```
+
+* param false: optional, disables peer certificate validation.
+
+## Complete example DTLS/CoAP client (Node 2)
+
+In this example the coap client is also the dtls client.
+The dlts client can connect to an coaps server which is listen on coaps port 5684, e.g to the [Node 1](#complete-example-for-dtlscoap-server-node-1) above.
+
+> Note: Node 1 and Node 2 must use the same mode. Either PSK or Certificate based.   
+
+```
+ Node 2
+---------
+|CoAPS  |
+|Client |--Connect to Server on Port 5684--> (Node 1)
+|       |
+---------
+```
+
+### with PSK
+
+```bash
+coaps start
+coaps set psk secretPSK Client_identity
+coaps connect 2001:620:190:ffa1::321
+coaps get time
+coaps disconnect
+```
+
+### with Certificate
+
+```bash
+coaps start (false)
+coaps set x509
+coaps connect <server_ip>
```

Comment:
The PSK method works for me.  However, the certificate method always results in `CoAP Secure not connected or disconnected` after this step.  Thoughts?

Reply:
The certificate verification not work without a CA chain. So it was necessary to disable the verification by cmd `coaps start false`. But i have added a CA yet to the CLI. Actually the certificate method also works with certificate verification between two thread nodes. I hope, it will also run by you, now.

Reply:
@ChrIgiSta, I tried this again without success.  I tried with Verify Peer Certificate on and off, both without success.  I end up with the same error `CoAP Secure not connected or disconnected.`

Reply:
@jwhui I can not reproduce this issue. Tested with commit fa6a4a5d8291699adeefb1c376c6615e108499b9.
Node Server CLI:
```
> channel
11
Done
> panid
ffff
Done
> panid 0x1234
Done
> ifconfig up
Done
> thread start
Done
> state
detached
Done
> state
leader
Done
> ipaddr
fdde:ad00:beef:0:0:ff:fe00:fc00
fdde:ad00:beef:0:0:ff:fe00:3c00
fdde:ad00:beef:0:5705:ad2a:6f2:32d1
fe80:0:0:0:b837:4b95:1329:d2bc
Done
> ping fe80:0:0:0:9c3f:a959:5b1a:678b
> 8 bytes from fe80:0:0:0:9c3f:a959:5b1a:678b: icmp_seq=2 hlim=64 time=8ms
> coaps start
Verify Peer Certificate: true. Coap Secure service started: Done
> coaps set x509
Coap Secure set own .X509 certificate: Done                                
> coaps resource test                                                      
Resource name is 'test': Done                                              
> CoAP Secure connected!                                                   
> Received coap secure request from [fe80:0:0:0:b837:4b95:1329:d2bc]: GET    No payload.
> coap secure response sent successfully!
```
Node Client CLI:
```
> channel
11
Done
> panid
ffff
Done
> panid 0x1234
Done
> ifconfig up
Done
> thread start
Done
> state
child
Done
> ipaddr
fdde:ad00:beef:0:0:ff:fe00:3c01
fdde:ad00:beef:0:66ae:5d97:15a5:35e9
fe80:0:0:0:9c3f:a959:5b1a:678b
Done
> ping fe80:0:0:0:b837:4b95:1329:d2bc
> 8 bytes from fe80:0:0:0:b837:4b95:1329:d2bc: icmp_seq=1 hlim=64 time=8ms
> coaps start
Verify Peer Certificate: true. Coap Secure service started: Done
> coaps set x509
Coap Secure set own .X509 certificate: Done
> coaps connect fe80:0:0:0:b837:4b95:1329:d2bc 
Coap Secure connect: Done
> CoAP Secure connected!
> coaps get test
Sending coap secure request: Done
> Received coap secure response
    CoapSecure RX Header Informations:
        Type 32 (ACK)
        Code 69 (Coap Code CONTENT)
    With payload [UTF8]:
helloWorld
```
Tested with two nRF52840 PDK's.
Which is your test environment? Eventually it can help to reproduce the issue.

Reply:
A further question. Do you test again our testserver (2001:620:190:ffa1::321)? 
If so, you must use the Port 5000. 
`coaps connect 2001:620:190:ffa1::321 5000`

Reply:
@ChrIgiSta, I was testing with the posix emulation platform.

Server Node:
```
> panid 0x1234
Done
> ifconfig up
Done
> thread start
Done
> state      
leader
Done
> coaps start
Verify Peer Certificate: true. Coap Secure service started: Done
> coaps set x509
Coap Secure set own .X509 certificate: Done
> coaps resource test
Resource name is 'test': Done
> ipaddr
fdde:ad00:beef:0:0:ff:fe00:fc00
fdde:ad00:beef:0:0:ff:fe00:7800
fdde:ad00:beef:0:3d61:cc9f:8022:cdfe
fe80:0:0:0:6c0a:28c1:fd0c:82cc
Done
> CoAP Secure not connected or disconnected.
```

Client Node:
```
> panid 0x1234
Done
> ifconfig up
Done
> thread start
Done
> state
child
Done
> coaps start
Verify Peer Certificate: true. Coap Secure service started: Done
> coaps set x509
Coap Secure set own .X509 certificate: Done
> coaps connect fe80:0:0:0:6c0a:28c1:fd0c:82cc
Coap Secure connect: Done
> CoAP Secure not connected or disconnected.
```

Reply:
@jwhui Thank you for explanation. I am be able to reproduce the issue with POSIX emulation. I figure out the problem as memory problem (heap). I push a fix today.

Reply:
@jwhui I've pushed a fix. Now it works as well with posix.

Reply:
@ChrIgiSta, with your latest changes, the example now works for me.  Thanks.